core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -2864,11 +2864,21 @@ public void refreshExtensions() throws ExtensionRefreshException {
             delta = ExtensionComponentSet.union(delta, ef.refresh().filtered());
         }
 
+        List<ExtensionList> listsToFireOnChangeListeners = new ArrayList<>();
         for (ExtensionList el : extensionLists.values()) {
-            el.refresh(delta);
+            if (el.refresh(delta)) {
+                listsToFireOnChangeListeners.add(el);
+            }
         }
         for (ExtensionList el : descriptorLists.values()) {
```

Comment:
Are `extensionLists` and `descriptorLists` mutually exclusive?

Reply:
The top-level elements in `extensionLists` and `descriptorLists` are distinct, but the inner elements of those lists are not. For example, running this in the script console:

```
def el = Jenkins.get().getExtensionList(org.jenkinsci.plugins.workflow.steps.StepDescriptor.class)
def dl = Jenkins.get().getDescriptorList(org.jenkinsci.plugins.workflow.steps.Step.class)
println(el.class.name + " " + System.identityHashCode(el))
println(dl.class.name + " " + System.identityHashCode(dl))
println(el.equals(dl))
```

Prints this:

```
hudson.ExtensionList 2011401743
hudson.DescriptorExtensionList 1431565206
true
```

Why do you ask?

Reply:
Oh I was looking at `listsToFireOnChangeListeners` being a `List` not a `Set` and was concerned about element overlap, but that does not matter here. Never mind.